src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java

```diff
@@ -157,19 +160,36 @@ private void streamToOutput() {
     }
 
     private void flushData(final Partition.PartitionKey pk, final SortedMap<Long, String> data) {
-        final String evt = EventStream.createStreamEvent(
-                pk.partition,
-                String.valueOf(offsets.get(pk).getSentOffset()),
-                new ArrayList<>(data.values()),
-                Optional.empty());
         try {
-            getOut().streamData(evt.getBytes(EventStream.UTF8));
+            final String offset = String.valueOf(offsets.get(pk).getSentOffset());
+            final String batch = serializeBatch(pk, offset, new ArrayList<>(data.values()));
+            getOut().streamData(batch.getBytes(EventStream.UTF8));
         } catch (final IOException e) {
             getLog().error("Failed to write data to output.", e);
             shutdownGracefully("Failed to write data to output");
         }
     }
 
+    private String serializeBatch(final Partition.PartitionKey partitionKey, final String offset,
+                                  final List<String> events) throws JsonProcessingException {
+
+        final String eventType = getContext().getEventTypesForTopics().get(partitionKey.topic);
+        final String token = getContext().getCursorTokenService().generateToken();
+        final SubscriptionCursor cursor = new SubscriptionCursor(partitionKey.partition, offset, eventType, token);
+        final String cursorSerialized = getContext().getObjectMapper().writeValueAsString(cursor);
+
+        final StringBuilder builder = new StringBuilder()
```

Comment:
No, it's not because of efficiency (actually on my benchmark jackson showed better results than making that by hand with string builder)

The thing is that the events are already serialized json strings, so we can't put events to some domain object and then serialize the object because the events will be serialized twice in that case. That's why we have to do this manual work with string builder. 
(Another possibility could be to deserialize events, put them to some batch-object and then serialize the whole object, but I'm afraid that will be very time-consuming)